internal/service/users.go

```diff
@@ -40,6 +45,49 @@ func (u *usersService) UpdateUserInfo(ctx context.Context, id string, input *Upd
 	return u.repo.Update(ctx, id, &upd)
 }
 
+func (u *usersService) Signup(ctx context.Context, input *SignupUserInput) error {
+	user, err := u.repo.GetByEmail(ctx, input.Email)
+	if err != nil && err != repository.ErrNotFound {
+		return err
+	}
+
+	if user != nil && user.Email == input.Email {
+		return ErrUserAlreadyExist
+	}
+
```

Comment:
For the second argument I suggest using bcrypt.DefaultCost (10) at least. This parameter has an exponential effect, so the difference is between MinCost (4) and the default one is huge.

This is a training project, and actual security means little. On the other hand, we are here to learn some proper practices, so we should not use cryptography parameters which are highly insecure.

The thing is, the purpose of bcrypt is the protection against brute-forcing password hashes (e.g. rainbow table attack). Setting the cost to minimum almost nullifies the effect of bcrypt (which is not high by modern standards, even on proper cost settings). As we are to use bcrypt according to the task definition, we should at least use a proper cost value.